app/presenters/tree_builder_timelines.rb

```diff
@@ -0,0 +1,80 @@
+class TreeBuilderTimelines < TreeBuilder
+  has_kids_for Hash, [:x_get_tree_hash_kids]
+
+  def override(node, object, _pid, _options)
+    node[:expand] = object[:expand] if object[:expand].present?
+  end
+
+  def initialize(name, type, sandbox, build = true, menu:)
+    @menu = menu
+    super(name, type, sandbox, build)
+  end
+
+  private
+
+  def set_locals_for_render
+    super.merge!(
+      :id_prefix  => 'timelines_',
+      :onclick    => "miqOnClickTimelineSelection",
+      :click_url  => "/dashboard/show_timeline/",
+      :tree_state => true
+    )
+  end
+
+  def tree_init_options(_tree_name)
+    {
+      :full_ids => true,
+      :lazy     => false,
+      :add_root => false
+    }
+  end
+
+  def root_options
+    {}
+  end
+
+  def x_get_tree_roots(count_only, _options)
+    menus = @menu.map do |item|
+      title = item.first
+      {
+        :text        => title,
+        :id          => "r__#{title}",
+        :icon        => 'pficon pficon-folder-close',
+        :tip         => title,
+        :cfmeNoClick => true,
+        :expand      => true,
+        :subsections => item.last
+      }
+    end
+    count_only_or_objects(count_only, menus.compact)
+  end
+
+  def x_get_tree_hash_kids(parent, count_only)
+    subsections = parent[:subsections].map do |item|
+      if item.kind_of?(Array)
+        title = item.first
+        {
+          :text        => title,
+          :id          => "p__#{title}",
+          :icon        => 'pficon pficon-folder-close',
+          :tip         => _("Group: %{:name}" % {:name => title}),
+          :expand      => false,
+          :cfmeNoClick => true,
+          :subsections => item.last
+        }
+      else
+        report = MiqReport.find_by(:name => item)
```

Comment:
@skateman @ZitaNemeckova So, I don't think I have time to get myself 100% up to speed, but think I can glean a bit from the changes in this file.  But, just so I make sure I am on the same page, here are the assumptions I am making:

1.  Looks like this is a shallow tree, and only goes about 1 level deep
2.  Of the report data you are getting, you only need the `.id` from the `.name`
3.  Report Name is unique (which is is, thankfully)

So, assuming that you might be able to do something like this:

```ruby
subsection_reports = MiqReport.where(:name => parent[:subsections].reject {|r| r.kind_of? String}).pluck(:name, :id).to_h
subsections = parent[:subsections].map do |item|
  if item.kind_of?(Array)
    # ...
  else
    if subsection_reports[item]
      {
        :text        => item,
        :id          => "#{subsection_reports[item]}__#{item}",
        :icon        => "fa fa-arrow-circle-o-right",
        :tip         => _("Report: %{:name}" % {:name => item}),
        :subsections => []
      }
    end
  end
```

This should change the possible `N+1` to a single call at the top of the method, also, it might only be one call if it is a shallow tree, but even if it is not, it should reduce the data returned and the number of queries made by a significant amount.